subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/DefaultTestExecuter.java

```diff
@@ -149,4 +162,61 @@ private int getMaxParallelForks(JvmTestExecutionSpec testExecutionSpec) {
         }
         return maxParallelForks;
     }
+
+    /**
+     * Create a classpath or modulePath, as a set of files, given both the files provided by the test spec and a set of
+     * modules to potentially load from the Gradle distribution. For each additional module to load, we check if the
+     * test files contains a jar which satisfies the module-to-load. If the test files do not contain a satisfactory
+     * jar, the given additional module is loaded from the Gradle distribution.
+     *
+     * TODO: This process is performed on a best-effort basis by matching against test file jar names. In the future
+     * may want to consider updating this code to take dependency-management into account, which can provide
+     * a more accurate view of {@code testFiles} that does not depend on string regex matching.
+     *
+     * @param testFiles A set of jars, as given from a {@link JvmTestExecutionSpec}'s classpath or modulePath.
+     * @param additionalModules The names of any additional modules to potentially load from the Gradle distribution.
+     *
+     * @return A set of files representing the constructed classpath or modulePath.
+     */
+    private Set<File> pathWithAdditionalJars(Iterable<? extends File> testFiles, Set<? extends TestFramework.DistributionModule> additionalModules) {
+        Set<File> outputFiles = new LinkedHashSet<File>();
+
+        Set<? extends TestFramework.DistributionModule> mutableAdditionalModules = new HashSet<TestFramework.DistributionModule>(additionalModules);
+        for (File f : testFiles) {
+            // Loop through each additional jar from the distribution. If the test classpath provides
+            // this jar, don't load it from the distribution.
+            Iterator<? extends TestFramework.DistributionModule> it = mutableAdditionalModules.iterator();
+            while (it.hasNext()) {
+                TestFramework.DistributionModule module = it.next();
+                if (module.getFileNameMatcher().matcher(f.getName()).matches()) {
+                    // The test classpath provides this jar.
+                    it.remove();
+                }
+            }
+            outputFiles.add(f);
+        }
+
+        Set<? extends TestFramework.DistributionModule> toLoad = mutableAdditionalModules;
+        int numMutableModules = mutableAdditionalModules.size();
+        if (numMutableModules > 0 && numMutableModules < additionalModules.size()) {
+            // The test classpath only provides some of the additional modules. So,
+            // load all of our modules so that we don't get version mismatches between our
+            // own dependencies. (Eg junit-platform-launcher-1.9.0 and junit-platform-commons-1.6.0)
+            toLoad = additionalModules;
```

Comment:
This is a really best-effort process that could get things wrong in a number of common scenarios - renamed jars, tree-shook jars, fat jars.  I think the fact we're doing this ought to be visible to the user.  Maybe as a warning, but at least as an info level message.

Reply:
I think a warning would be too verbose here. We can encounter these duplicate classes any time a user uses the default test suite or standalone test tasks while also declaring test framework dependencies. 

Its tricky, since with test suites users don't need to declare their dependencies, but without test suites, if users want to select their own versions of test framework dependencies, they need to declare them explicitly. 

I think nagging the user at this time without de-incubating test suites is premature.